hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/ConsistentBucketIndexBulkInsertPartitionerWithRows.java

```diff
@@ -105,10 +121,55 @@ public int numPartitions() {
       }
     };
 
-    return rows.sparkSession().createDataFrame(rowJavaRDD
-        .mapToPair(row -> new Tuple2<>(getBucketId(row), row))
-        .partitionBy(partitioner)
-        .values(), rows.schema());
+    if (sortColumnNames != null && sortColumnNames.length > 0) {
+      return rows.sparkSession().createDataFrame(rowJavaRDD
+              .mapToPair(row -> new Tuple2<>(row, row))
```

Comment:
We will still need the row for comparing and sort it, so keep this line `  .mapToPair(row -> new Tuple2<>(row, row))` is OK. 

Also comparing with partitionBy + sortWithinPartitions, repartitionAndSortWithinPartitions will be more efficient because it performs the shuffle operation only once, with both repartitioning and sorting happening in the same step. What do you think?

Reply:
> For the second issue, partitionBy + sortWithinPartitions also does shuffle only once?

Yes, I was mistaken. I've made the changes that use spark native API and remove custom comparators. PTAL~